src/QueryEditor.tsx

```diff
@@ -1,48 +1,269 @@
+        if (!this.loadingAssets && this.dataSourceId != this.props.datasource.id) {
+            this.loadingAssets = true;
+            if (this.dataSourceId != this.props.datasource.id && this.initialLoadingComplete) {
+                this.resetAllValues();
+                this.dataSourceId = this.props.datasource.id;
+            }
+
+            this.props.datasource.getAssets().then(
+                (assets: Asset[]) => {
+                    const assetSelectOptions: Array<SelectableValue<string>> = [];
+                    for (const asset of assets) {
+                        assetSelectOptions.push({label: asset.name, value: asset.id, asset: asset});
+                    }
+                    this.setState(prevState => ({
+                        ...prevState,
+                        assetOptions: assetSelectOptions,
+                    }));
+                    this.loadingAssets = false;
+                    this.initialLoadingComplete = true;
+                    // initial render does not update the select loading state, hence the force update
+                    // it won't trigger a re-rendering of the view, since the above checks prevent this
+                    this.forceUpdate();
+                },
+                // in case an error is thrown, stop the loading animation
+                () => {
+                    this.loadingAssets = false;
+                }
+            );
+        }
+    }
+
+    private loadTopics(assetId: string): void {
+        this.loadingTopics = true;
+        this.props.datasource.getTopics(assetId).then(
+            (topics: string[]) => {
+                let topicsSelectOptions: Array<SelectableValue<string>> = [];
+                for (const topic of topics) {
+                    topicsSelectOptions.push({label: topic, value: topic});
+                }
+                this.loadingTopics = false;
+                this.setState(prevState => ({
+                    ...prevState,
+                    topicOptions: topicsSelectOptions,
+                }));
+                if (topicsSelectOptions.length === 0) {
+                    // if no topics were found, reset dataKey and topic values options
+                    this.resetTopicAndDataKeyValues();
+                }
+            },
+            // in case an error is thrown, stop the loading animation
+            () => {
+                this.loadingTopics = false;
+            }
+        )
+    }
+
+    private loadKeys(assetId: string, topic: string): void {
+        this.loadingDataKeys = true;
+        this.props.datasource.getKeys(assetId, topic).then(
+            (keys: string[]) => {
+                let keySelectOptions: Array<SelectableValue<string>> = [];
+                for (const key of keys) {
+                    keySelectOptions.push({label: key, value: key});
+                }
+                this.loadingDataKeys = false;
```

Comment:
store this in state?

Reply:
I had that at one point but it bloats the function more, since I have to merge with the previous state in order not to lose all other propteries in it 🤔

Reply:
ok, makes sense as it should never trigger a rerender, so it is fine to store it in **this**